### PR TITLE
Give the possibility to update an existing card by modifying only some properties (#6456)

### LINKED ISSUE #6456: Give the possibility to update an existing card by modifying only some properties
1. And to take into account the both CardActionEnum during the updating :
 1.1. `KEEP_EXISTING_PUBLISH_DATE`,
 1.2. `KEEP_EXISTING_ACKS_AND_READS`
2. Allow to choose the properties to be modified

### PR DESCRIPTION
**Pull request by vlo-rte**
Sonar is complaining but I think we can set this sonar issue as "accepted" for the same reason than following : 

image

- In release note :
  -  In chapter :  Features
  -  Text : #6456 : Give the possibility to update an existing card by modifying only some properties

### CODE REVIEW

File: src/test/api/karate/cards/patchCard.feature [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    And request cardPatch
+    When method patch
+    Then status 404
+
+# Patch card
+    Given url opfabPublishCardUrl + 'cards/' + cardId
+    And header Authorization = 'Bearer ' + authToken
+    And request cardPatch
+    When method patch
+    Then status 200
```
Comment (freddidierRTE) [inline src/test/api/karate/cards/patchCard.feature:91]: Check card has been update via a GET

File: src/test/api/karate/cards/patchCard.feature [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	"title" : {"key" : "defaultProcess.title"},
+	"data" : {"message":"a message"}
+}
+"""
+
+    * def cardPatch =
+"""
+{
+	"severity" : "COMPLIANT"
+}
```
Comment (freddidierRTE) [inline src/test/api/karate/cards/patchCard.feature]: Patch data field + entityRecipients as well to have a more complete test

Check for example startDate , title and publisher has not been changed

File: services/cards-publication/src/test/java/org/opfab/cards/publication/services/CardProcessServiceShould.java [hunk 3]
```diff
@@ -284,6 +284,35 @@ void GIVEN_a_publisher_WHEN_sending_5_cards_THEN_cards_are_saved() {
...
+        void GIVEN_a_publisher_WHEN_patching_a_card_THEN_card_is_saved() {
+                Card initialCard = generateOneCard();
+                cardProcessingService.processCard(initialCard);
+                Assertions.assertThat(checkCardCount(1)).isTrue();
+                Assertions.assertThat(checkArchiveCount(1)).isTrue();
+
+                Card patchCard = Card.builder().severity(SeverityEnum.ALARM).build();
+                cardProcessingService.patchCard(initialCard.getId(), patchCard, Optional.empty(), Optional.empty());
+                Assertions.assertThat(checkCardCount(1)).isTrue();
+                Assertions.assertThat(checkArchiveCount(2)).isTrue();
```
Comment (freddidierRTE) [inline services/cards-publication/src/test/java/org/opfab/cards/publication/services/CardProcessServiceShould.java:306]: check severity  is updated

File: services/cards-publication/src/main/java/org/opfab/cards/publication/services/CardValidationService.java [hunk 4]
```diff
@@ -117,14 +117,25 @@ void validate(Card c) throws ConstraintViolationException {
                     "constraint violation : forbidden characters ('#','?','/') in process or processInstanceId", null);
     }
 
+    void validateCardForPatch(Card cardForPatch, Card initialCard) throws ConstraintViolationException {
+        if ((cardForPatch.getProcess() != null) && !cardForPatch.getProcess().equals(initialCard.getProcess()))
+            throw new ConstraintViolationException("The current process field " + initialCard.getProcess()
+                    + " can not be patched with new value " + cardForPatch.getProcess(), null);
```
Comment (freddidierRTE) [inline services/cards-publication/src/main/java/org/opfab/cards/publication/services/CardValidationService.java:123]: Proposal : 

```
String.format("The current process field %s cannot be patched with new value %s",initialCard.getProcess(),  cardForPatch.getProcess())
``` 
```

File: services/cards-publication/src/main/java/org/opfab/cards/publication/model/Card.java [hunk 5]
```diff
@@ -170,6 +171,97 @@ public void prepare(Instant publishDate) {
...
+
+        result.publisher = ObjectUtils.getNotNullOrDefault(other.getPublisher(), this.getPublisher());
+
+        result.processVersion = ObjectUtils.getNotNullOrDefault(other.getProcessVersion(),
+                this.getProcessVersion());
+
+        result.process = this.process;
+        result.processInstanceId = this.processInstanceId;
+
+        result.state = ObjectUtils.getNotNullOrDefault(other.getState(),
```
Comment (freddidierRTE) [inline services/cards-publication/src/main/java/org/opfab/cards/publication/model/Card.java:191]: proposal : 
```
Optional.ofNullable(other.state).orElse(this.state);
```

Comment (vlo-rte) [reply services/cards-publication/src/main/java/org/opfab/cards/publication/model/Card.java:191]: After discussion, we've decided to keep it as it is.

File: services/cards-publication/src/main/java/org/opfab/cards/publication/services/CardValidationService.java (hunk 4 above)
Comment (vlo-rte) [reply services/cards-publication/src/main/java/org/opfab/cards/publication/services/CardValidationService.java:123]: After discussion, we've decided to keep it as it is.

File: services/cards-publication/src/test/java/org/opfab/cards/publication/services/CardProcessServiceShould.java (hunk 3 above)
Comment (vlo-rte) [reply services/cards-publication/src/test/java/org/opfab/cards/publication/services/CardProcessServiceShould.java:306]: Done.

File: src/test/api/karate/cards/patchCard.feature (hunk 2 above)
Comment (vlo-rte) [reply src/test/api/karate/cards/patchCard.feature]: Done.

File: src/test/api/karate/cards/patchCard.feature (hunk 1 above)
Comment (vlo-rte) [reply src/test/api/karate/cards/patchCard.feature:91]: Done.

**Review by freddidierRTE — APPROVED** [review]

**Pull request merged**